src/udiff/_diff_array.py

```diff
@@ -1,139 +1,117 @@
+        try:
+            if func is np.ufunc.__call__:
+                vjpmaker = primitive_vjps[args[0]]
+            else:
+                vjpmaker = primitive_vjps[func]
+        except KeyError:
+            raise NotImplementedError("VJP of func not defined")
+        parent_argnums = []
+        vjp_args = []
+        pn = 0
+        for arg in args:
+            if isinstance(arg, DiffArray):
+                self._parents.append((pn, arg))
+                parent_argnums.append(pn)
+                pn += 1
+                vjp_args.append(arg.value)
+
+        self._vjp = vjpmaker(tuple(parent_argnums), self.value, tuple(vjp_args), kwargs)
 
     def __str__(self):
-        return "<{}, name={}, arr=\n{}\n>".format(
+        return "<{}, name={}, value=\n{}\n>".format(
             type(self).__name__,
-            repr(self.var.name) if self.var is not None else "unbound",
-            str(self.arr),
+            repr(self.name) if self.name is not None else "unbound",
+            str(self.value),
         )
 
-    __repr__ = __str__
-    __hash__ = object.__hash__
-
-    def __setitem__(self, k, v):
-        if not isinstance(v, type(self)):
-            raise ValueError("v must be of the same type as self")
-
-        if not all(i in v._diffs.keys() for i in self._diffs.keys()):
-            raise ValueError("v doesn't have all required diffs")
-
-        self._arr[k] = v._arr
-
-        for i in self._diffs:
-            self._diffs[i][k] = v._diffs[i]
-
-
-class Variable:
-    def __init__(self, name):
-        if not isinstance(name, str):
-            raise ValueError("k must be a string")
-
-        self._name = str(name)
-
-    @property
-    def name(self):
-        return self._name
-
-    def __str__(self):
-        return "Variable({})".format(repr(self._name))
+    def backward(self, grad_variables=None, end_node=None):
+        """
+        Backpropagation.
+        Traverse computation graph backwards in topological order from the end node.
+        For each node, compute local gradient contribution and accumulate.
+        """
+        from ._uarray_plug import NoGradBackend
+
+        with ua.set_backend(NoGradBackend()):
+            if grad_variables is None:
+                grad_variables = np.ones(self.value.shape)
+            if end_node is None:
+                end_node = self.name
+            if self._diff is None or self._visit != end_node:
+                self._diff = np.zeros(self.value.shape)
+            self._diff += grad_variables
+            self._visit = end_node
+            if self._vjp:
+                diffs = self._vjp(grad_variables)
+                for pn, p in self._parents:
+                    p.backward(diffs[pn], self._visit)
+
+    def backward_jacobian(self):
+        """Backpropagation.
+        Traverse computation graph backwards in topological order from the end node.
+        For each node, compute local Jacobian contribution and accumulate.
+        """
+        pass
```

Comment:
Maybe raise `NotImplementedError` here.
```suggestion
        raise NotImplementedError
```